File: python/src/problems101_200/binary_tree_level_order_traversal.py
from abc import ABC, abstractmethod
from collections import deque
from typing import override

from src.concepts.tree_node import TreeNode


class Problem(ABC):
    @abstractmethod
    def level_order(self, root: TreeNode | None) -> list[list[int]]:
        pass


class Default(Problem):
    @override
    def level_order(self, root: TreeNode | None) -> list[list[int]]:
        lists: list[list[int]] = []
        if not root:
            return lists

        queue: deque[TreeNode] = deque([root])
        while len(queue) > 0:
            size: int = len(queue)
            level: list[int] = []
            for _ in range(size):
                current = queue.popleft()
                level.append(current.val)
                if current.left:
                    queue.append(current.left)
                if current.right:
                    queue.append(current.right)
            lists.append(level)
        return lists


SOLUTIONS: list[Problem] = [Default()]

File: javascript/src/problems101_200/binary-tree-level-order-traversal.js
class Problem {
  levelOrder(root) {
    throw new Error(`Not implemented (${root}).`);
  }
}

class Default extends Problem {
  levelOrder(root) {
    const lists = [];
    if (!root) {
      return lists;
    }

    const queue = [root];
    while (queue.length > 0) {
      const size = queue.length;
      const level = [];
      for (let i = 0; i < size; i++) {
        const current = queue.shift();
        level.push(current.val);
        if (current.left) {
          queue.push(current.left);
        }
        if (current.right) {
          queue.push(current.right);
        }
      }
      lists.push(level);
    }
    return lists;
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;
